File: android/app/src/main/java/com/cactus/remoteterminal/notify/Notifier.kt
package com.cactus.remoteterminal.notify

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import com.cactus.remoteterminal.R
import com.cactus.remoteterminal.data.AgentRepository
import com.cactus.remoteterminal.data.SessionRepository
import com.cactus.remoteterminal.data.Settings
import com.cactus.remoteterminal.data.TerminalSession
import com.cactus.remoteterminal.net.RelayClient
import com.cactus.remoteterminal.protocol.RelayEvent
import com.cactus.remoteterminal.ui.MainActivity

/**
 * Optional, quiet notifications: a machine you are using went offline, a
 * terminal's process exited, a terminal rang the bell. Only raised while the
 * app is in the background and only when the corresponding setting is on;
 * never for ordinary output.
 */
class Notifier(
    private val context: Context,
    private val settings: Settings,
    private val client: RelayClient,
    private val agents: AgentRepository,
    private val sessions: SessionRepository,
) : RelayClient.Listener {

    @Volatile var foreground = true
    private val usedAgents = HashSet<String>()

    init {
        createChannels()
        client.addListener(this)
        sessions.onSessionExited = { s -> if (!foreground && settings.notifyExit) notifyExit(s) }
        sessions.onBell = { s ->
            if (foreground && settings.bell == "vibrate") vibrate()
            if (!foreground && settings.notifyBell) notifyBell(s)
        }
    }

    private var lastVibrate = 0L
    private fun vibrate() {
        val now = System.currentTimeMillis()
        if (now - lastVibrate < 300) return // a flood of bells should not buzz continuously
        lastVibrate = now
        try {
            val v = if (Build.VERSION.SDK_INT >= 31) {
                (context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as android.os.VibratorManager).defaultVibrator
            } else @Suppress("DEPRECATION") (context.getSystemService(Context.VIBRATOR_SERVICE) as android.os.Vibrator)
            if (Build.VERSION.SDK_INT >= 26) v.vibrate(android.os.VibrationEffect.createOneShot(40, android.os.VibrationEffect.DEFAULT_AMPLITUDE))
            else @Suppress("DEPRECATION") v.vibrate(40)
        } catch (_: Throwable) { /* no vibrator */ }
    }

    /** Remember machines the user actively opened, so offline alerts are only for those. */
    fun noteAgentUsed(agentId: String) { usedAgents.add(agentId) }

    override fun onRelayEvent(event: RelayEvent) {
        if (event !is RelayEvent.AgentOffline) return
        // Two ways to opt in: the global "machine goes offline" setting for
        // machines you have opened, or per-machine connection alerts.
        val perMachine = settings.connectionAlerts(event.agentId)
        val global = settings.notifyAgentOffline && event.agentId in usedAgents
        if (!perMachine && !(global && !foreground)) return
        if (foreground && !perMachine) return
        val name = agents.agent(event.agentId)?.name ?: "A machine"
        post(
            CHANNEL_STATUS, event.agentId.hashCode(),
            context.getString(R.string.notif_offline_title, name),
            context.getString(R.string.notif_offline_text),
        )
    }

    private fun notifyExit(s: TerminalSession) {
        val machine = agents.agent(s.agentId)?.name ?: ""
        post(CHANNEL_TERMINAL, s.sessionId.hashCode(), context.getString(R.string.notif_exit_title, s.displayTitle),
            context.getString(R.string.notif_exit_text, machine, s.exitCode ?: 0), s)
    }

    private fun notifyBell(s: TerminalSession) {
        val machine = agents.agent(s.agentId)?.name ?: ""
        post(CHANNEL_TERMINAL, s.sessionId.hashCode(), context.getString(R.string.notif_bell_title, s.displayTitle), machine, s)
    }

    private fun post(channel: String, id: Int, title: String, text: String, session: TerminalSession? = null) {
        if (Build.VERSION.SDK_INT >= 33 &&
            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) return
        val intent = Intent(context, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP
            if (session != null) { putExtra(MainActivity.EXTRA_AGENT, session.agentId); putExtra(MainActivity.EXTRA_SESSION, session.sessionId) }
        }
        val pi = PendingIntent.getActivity(context, id, intent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
        val n = NotificationCompat.Builder(context, channel)
            .setSmallIcon(R.drawable.ic_terminal)
            .setContentTitle(title)
            .setContentText(text)
            .setContentIntent(pi)
            .setAutoCancel(true)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .build()
        try { NotificationManagerCompat.from(context).notify(id, n) } catch (_: SecurityException) { /* permission revoked */ }
    }

    private fun createChannels() {
        if (Build.VERSION.SDK_INT < 26) return
        val nm = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        nm.createNotificationChannel(NotificationChannel(CHANNEL_STATUS, context.getString(R.string.channel_status), NotificationManager.IMPORTANCE_DEFAULT).apply {
            description = context.getString(R.string.channel_status_desc)
        })
        nm.createNotificationChannel(NotificationChannel(CHANNEL_TERMINAL, context.getString(R.string.channel_terminal), NotificationManager.IMPORTANCE_DEFAULT).apply {
            description = context.getString(R.string.channel_terminal_desc)
        })
    }

    companion object {
        const val CHANNEL_STATUS = "status"
        const val CHANNEL_TERMINAL = "terminal"
    }
}

File: desktop/ui/js/core/notifier.js
/**
 * Optional, quiet notifications — a port of `Notifier.kt`: a machine you are
 * using went offline, a terminal's process exited, a terminal rang the bell.
 * Raised only while the window is not in front and only when the corresponding
 * setting is on; never for ordinary output.
 */

import { notify } from './platform.js';
import { S } from '../ui/strings.js';

export class Notifier {
  constructor(settings, client, agents, sessions) {
    this.settings = settings;
    this.client = client;
    this.agents = agents;
    this.sessions = sessions;
    this.foreground = true;
    this.usedAgents = new Set();
    this.lastBell = 0;

    client.on('event', (event) => this.onEvent(event));

    sessions.onSessionExited = (s) => {
      if (!this.foreground && settings.notifyOnFinish(s.key)) this.notifyExit(s);
    };
    sessions.onBell = (s) => {
      if (this.foreground && settings.bell === 'sound') this.beep();
      if (!this.foreground && settings.notifyBell) this.notifyBell(s);
    };
  }

  /** Remember machines the user actively opened, so offline alerts are only for those. */
  noteAgentUsed(agentId) { this.usedAgents.add(agentId); }

  onEvent(event) {
    if (event.kind !== 'agentOffline') return;
    // Two ways to opt in: the global "machine goes offline" setting for
    // machines you have opened, or per-machine connection alerts.
    const perMachine = this.settings.connectionAlerts(event.agentId);
    const global = this.settings.notifyAgentOffline && this.usedAgents.has(event.agentId);
    if (!perMachine && !(global && !this.foreground)) return;
    if (this.foreground && !perMachine) return;
    const name = this.agents.agent(event.agentId)?.name || 'A machine';
    notify(S.notifOfflineTitle(name), S.notifOfflineText);
  }

  notifyExit(s) {
    const machine = this.agents.agent(s.agentId)?.name ?? '';
    notify(S.notifExitTitle(s.displayTitle), S.notifExitText(machine, s.exitCode ?? 0));
  }

  notifyBell(s) {
    const machine = this.agents.agent(s.agentId)?.name ?? '';
    notify(S.notifBellTitle(s.displayTitle), machine);
  }

  /**
   * The desktop's answer to the phone's vibrate: a short tone through the Web
   * Audio API, so no sound file has to ship. A flood of bells should not become
   * a flood of beeps.
   */
  beep() {
    const now = Date.now();
    if (now - this.lastBell < 300) return;
    this.lastBell = now;
    try {
      const Ctx = window.AudioContext || window.webkitAudioContext;
      if (!Ctx) return;
      this.audio = this.audio || new Ctx();
      const ctx = this.audio;
      const osc = ctx.createOscillator();
      const gain = ctx.createGain();
      osc.type = 'sine';
      osc.frequency.value = 880;
      gain.gain.setValueAtTime(0.0001, ctx.currentTime);
      gain.gain.exponentialRampToValueAtTime(0.08, ctx.currentTime + 0.01);
      gain.gain.exponentialRampToValueAtTime(0.0001, ctx.currentTime + 0.12);
      osc.connect(gain).connect(ctx.destination);
      osc.start();
      osc.stop(ctx.currentTime + 0.13);
    } catch {
      // No audio device, or the context is blocked: a silent bell is fine.
    }
  }
}
